### ISSUE TITLE
[bug] `prefer-use-state-lazy-initialization`: false positives with`new Set()`, `new Map()` etc

### ISSUE THREAD

**Issue by silverwind** [Rel1cx/eslint-react#1606]
The rule flags these and wants a `() =>` wrapper:

```tsx
useState(new Set()); // error
useState(new Map()); // error
useState(new Date()); // error
```

I suggest at minimum ignoring common built-in constructors like `Set`, `Map`, `Date`, `WeakSet`, `WeakMap` etc. with zero arguments, they are cheap and likely cheaper than adding a function wrapper.

**Comment by SukkaW** [comment]
> they are cheap and likely cheaper than adding a function wrapper.

For v8-based browsers like Chromium series and JavaScriptCore-based browsers like Safari, yes.

For Spidermonkey-based browsers like Firefox (and its forks), no.

SpiderMonkey is notorious for its very slow GC. If you allocate many hanging objects, it may take SpiderMonkey up to 100ms to GC them. That's blocking the main event loop for 100ms, and dropping 7 frames.

Learned that lesson the hard way with only 200 temp `Map` objects (triggered a "Major GC" which took Firefox 130ms).

**Comment by silverwind** [comment]
Benchmarked that with 10M iterations:

| Pattern | JSC | SpiderMonkey | V8 |
|---|---|---|---|
| `new Set()` (no lazy) | 40ms | **981ms** | 90ms |
| `() => new Set()` (lazy) | 9ms | 24ms | 28ms |

Commands used:

```bash
eshost -e '(()=>{let d,s=Date.now();for(let i=0;i<1e7;i++)d=new Set();return Date.now()-s+"ms "+d})()'
eshost -e '(()=>{let d,s=Date.now();for(let i=0;i<1e7;i++)d=()=>new Set();return Date.now()-s+"ms "+d})()'
```

There is a difference, yes. But I think it's very unlikely that a typical react app would render 10M times, so it seems like a unneccesary micro-optimization.

**Comment by silverwind** [comment]
My eshost engines were out of date (2023), I updated to latest and SM appears to be the fastest now in the non-wrapped case:

| Pattern | JSC | SM (old) | SM (new) | V8 (old) | V8 (new) |
|---|---|---|---|---|---|
| `new Set()` | 39ms | **981ms** | **31ms** | 90ms | 78ms |
| `() => new Set()` | 4ms | 24ms | 22ms | 28ms | 18ms |

Wrapped is faster, but ultimately this difference will not matter in React apps, at least not with empty sets. For populated ones, it's likely better to wrap. A argument could be made to always wrap for consistency.

**Comment by SukkaW** [comment]
> Wrapped is faster, but ultimately this difference will not matter in React apps, at least not with empty sets. For populated ones, it's likely better to wrap. A argument could be made to always wrap for consistency.

Creation is actually not a huge problem. GC is the issue. SpiderMonkey's GC is ridiculously slow. For `useState(new Map())`, it means every time your component renders, you create a temp never-used object, which adds up GC pressure. As I said before, it once took 130ms for Firefox to do a major GC, at the cost of blocking the main event loop for 9 frames.

**Comment by silverwind** [comment]
Closing as I think consistency wins and using a mix of non-function vs. function would ultimately be worse.

**Comment by silverwind** [comment]
> Creation is actually not a huge problem. GC is the issue. SpiderMonkey's GC is ridiculously slow. For `useState(new Map())`, it means every time your component renders, you create a temp never-used object, which adds up GC pressure. As I said before, it once took 130ms for Firefox to do a major GC, at the cost of blocking the main event loop for 9 frames.

I did some non-scientific GC benchmarks too and did not notice SM being much slower than V8 or JSC. So maybe it has already caught up.
